soroban-env-host/src/host/metered_xdr.rs

```diff
@@ -0,0 +1,59 @@
+use std::io::Write;
+
+use crate::{
+    budget::CostType,
+    xdr::{ReadXdr, WriteXdr},
+    Host, HostError,
+};
+
+struct MeteredWrite<'a, W: Write> {
+    host: &'a Host,
+    w: &'a mut W,
+}
+
+impl<'a, W> Write for MeteredWrite<'a, W>
+where
+    W: Write,
+{
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.host
+            .charge_budget(CostType::ValSer, buf.len() as u64)
+            .map_err(|e| Into::<std::io::Error>::into(e))?;
+        self.w.write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.w.flush()
+    }
+}
+
+impl Host {
+    pub(crate) fn metered_write_xdr(
+        &self,
+        obj: &impl WriteXdr,
+        w: &mut Vec<u8>,
+    ) -> Result<(), HostError> {
+        let mut w = MeteredWrite { host: &self, w };
+        obj.write_xdr(&mut w).map_err(|e| match e {
+            // The `MeteredWrite` may result in a HostError stored in `io::Error`,
+            // which is wrapped in `xdr::Error`. Here we fish `HostError` back out.
+            soroban_env_common::xdr::Error::Io(e1) => {
+                if let Some(e2) = e1.into_inner() {
+                    if let Some(e3) = (*e2).downcast_ref::<HostError>() {
+                        e3.clone()
+                    } else {
+                        self.err_general("failed to write xdr")
+                    }
+                } else {
+                    self.err_general("failed to write xdr")
+                }
+            }
+            _ => self.err_general("failed to write xdr"),
+        })
```

Comment:
Oh this is nice. Fyi you can use the common `.source()` function on the Error rather than needing to match and get inner:
```suggestion
        obj.write_xdr(&mut w).map_err(|e| {
            if let Some(e2) = e.source() {
                if let Some(e3) = (*e2).downcast_ref::<HostError>() {
                    e3.clone()
                } else {
                    self.err_general("failed to write xdr")
                }
            } else {
                self.err_general("failed to write xdr")
            }
        })
```